app/src/lib/markdown-filters/emoji-filter.ts

```diff
@@ -0,0 +1,138 @@
+   * emoji image nodes using base64 data uri where emoji references are.
+   *
+   * Example: A text node of "That is great! :+1: Good Job!" Becomes three
+   * nodes: ["That is great! ",<img src="data uri for :+1:>, " Good Job!"]
+   */
+  public async filter(node: Node): Promise<ReadonlyArray<Node> | null> {
+    if (!(node instanceof Text)) {
+      fatalError(
+        'Emoji filter requires text nodes; otherwise we may inadvertently replace non text elements.'
+      )
+    }
+
+    if (node.textContent === null || !node.textContent.includes(':')) {
+      return null
+    }
+
+    let text = node.textContent
+    const emojiMatches = text.match(this.emojiRegex)
+    if (emojiMatches === null) {
+      return null
+    }
+
+    const nodes: Array<Text | HTMLImageElement> = []
+    for (let i = 0; i < emojiMatches.length; i++) {
+      const emojiKey = emojiMatches[i]
+      const emojiPath = this.emojiFilePath.get(emojiKey)
+      if (emojiPath === undefined) {
+        continue
+      }
+
+      const emojiPosition = text.indexOf(emojiMatches[0])
+      const textBeforeEmoji = text.slice(0, emojiPosition)
+      const textNodeBeforeEmoji = document.createTextNode(textBeforeEmoji)
+      nodes.push(textNodeBeforeEmoji)
+
+      const emojiImg = await this.createEmojiNode(emojiPath)
+      nodes.push(emojiImg)
+
+      text = text.slice(emojiPosition + emojiKey.length)
+    }
+
+    if (text !== '') {
+      const trailingTextNode = document.createTextNode(text)
+      nodes.push(trailingTextNode)
+    }
+
+    return nodes
+  }
+
+  /**
+   * Method to build an emoji image node to insert in place of the emoji ref
+   */
+  private async createEmojiNode(emojiPath: string) {
+    const dataURI = await this.getBase64FromImageUrl(emojiPath)
+    const emojiImg = new Image()
+    emojiImg.classList.add('emoji')
+    emojiImg.src = dataURI
+    return emojiImg
+  }
+
+  /**
+   * Method to obtain an images base 64 data uri from it's file path.
+   * - It checks cache, if not, reads from file, then stores in cache.
+   */
+  private async getBase64FromImageUrl(filePath: string): Promise<string> {
+    const cached = this.emojiBase64URICache.get(filePath)
+    if (cached !== undefined) {
+      return cached
+    }
+    const imageBuffer = await FSE.readFile(uri2path(filePath))
+    const b64src = imageBuffer.toString('base64')
+    const uri = `data:image/png;base64,${b64src}`
+    this.emojiBase64URICache.set(filePath, uri)
+
+    return uri
+  }
+
+  /**
+   * Builds a regular expression that is looking for all group of characters
+   * that represents any emoji ref (or map key) in the provided map.
+   *
+   * @param emoji Map from the emoji ref (e.g., :+1:) to the image's local path.
+   */
+  private buildEmojiRegExp(emoji: Map<string, string>): RegExp {
+    const emojiGroups = [...emoji.keys()]
+      .map(emoji => escapeRegExp(emoji))
+      .join('|')
+      .slice(0, -1)
```

Comment:
This would remove the last `:` from the last emoji and would cause us to match `:zzz` instead of `:zzz:`. I suggest we remove it

```suggestion
```